cassdegrees/api/degreeRules.py

```diff
@@ -1,52 +1,161 @@
+
         return tree
-        #create rules, return rules. Then take a list of rules in a bool alg expression and build the json
 
-    # Add rule as another AND?
+    def append_rule(self, rule):
+        """
+        Takes a rule that should be independent of the existing rules for a degree
+        and appends it to the degree rules with an AND operator.
+
+        :param rule: (Dict) the rule to be appended
+        """
+
+        if self.expressionTree[0] == "AND":
+            self.expressionTree.append(rule)
+        else:
+            self.expressionTree = ["AND", self.expressionTree, rule]
+
+    def get_rules_as_json(self):
+        """
+        :return: a JSON serialised string representing the rules expression tree
+        """
+
+        return json.dumps(self.expressionTree)
+
+    """
+        DEGREE RULES
+            Functions that return dictionary representations of rules
+    """
+
+    @staticmethod
+    def rule_course_selection(code):
+        """
+        Creates a rule which lists a single course code which
+        must be completed for a degree.
+
+        :param code: (String) the course code
+        :return: (Dict) a course selection rule
+        """
+
+        rule = {"type": "course selection", "code": code}
+        return rule
+
+    @staticmethod
+    def rule_subplan_selection(code):
+        """
+        Creates a rule which lists a single subplan code which
+        must be completed for a degree.
+
+        :param code: (String) the subplan code
+        :return: (Dict) a subplan selection rule
+        """
+
+        rule = {"type": "subplan selection", "code": code}
+        return rule
+
+    @staticmethod
+    def rule_electives(units):
+        """
+        Creates a rule which describes a number of units worth
+        of elective courses which must be completed for the degree.
+
+        :param units: (Int) number of units of available electives
+        :return: (Dict) an elective rule
+        """
+
+        rule = {"type": "electives", "units": units}
+        return rule
+
+    @staticmethod
+    def rule_subject_requirement(subject, level, minimum, units):
+        """
+        Creates a rule which describes a number of units which must
+        be completed, where each course is from a specific subject area
+        and of a given level (or potentially above).
+
+        :param subject: (String) a 4 letter course code i.e "ARTS", "COMP" etc.
+        :param level: (Int) a number in the range [1000, 2000, ... , 9000]
+        :param minimum: (Boolean) true = courses of the level or above count towards the rule,
+                                  false = only course equal to the level count towards the rule
+        :param units: (Int) number of units that must be completed from the subject area and level
+        :return: (Dict) a subject requirement rule
+        """
+
+        rule = {"type": "subject requirement", "subject": subject, "level": level, "minimum": minimum, "units": units}
```

Comment:
This is useful for things like "No more than 24 units of 1000-level ARTS courses", but we may also want to represent restrictions/enforcements on the number of ARTS courses exclusively, as well as enforcements/restrictions on the number of year/level-specific courses exclusively. Function below has the same issue

Reply:
I just wasn’t sure if this represented a finalised version. You’re probably right in thinking that we should update as we go so I’ll re-check my previous comments tomorrow (well today I guess) to see if they are fixed and will give you the go-ahead if they are